### PR TITLE
Wip scrub repair

### PR DESCRIPTION
**Pull request by ukernel**
1. make scrub stack code queue CInode instead of CDentry
2. make scrub stack code be able repair incorrect fragstat/neststat (enabled by option of scrub_path asok commad)

### CODE REVIEW

**Comment by ukernel** [comment]
https://github.com/ceph/ceph-qa-suite/pull/731

**Comment by dachary** [comment]
@ukernel please ignore the bot failure, it is a false negative (http://tracker.ceph.com/issues/14092).

**Comment by gregsfortytwo** [comment]
"mds: skip scrubbing dentries not modified since last scrub" needs to be optional (or perhaps based on time since the scrub stamp was last updated) — part of the point of scrubbing is that it actually *checks* the data is good.

File: src/mds/CDir.cc [hunk 1]
```diff
@@ -2974,6 +2986,13 @@ int CDir::_next_dentry_on_set(set<dentry_key_t>& dns, bool missing_okay,
     // okay, we got a  dentry
     dns.erase(dnkey);
 
+    if (dn->get_projected_version() < scrub_infop->last_recursive.version) {
+      dout(15) << " skip dentry " << dnkey.name
+	       << ", no change since last scrub" << dendl;
+      continue;
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: This needs to be optional (or perhaps based on time since the scrub stamp was last updated) — part of the point of scrubbing is that it actually checks the data is good.

File: src/mds/CDir.cc [hunk 2]
```diff
@@ -3061,13 +3081,17 @@ void CDir::scrub_maybe_delete_info()
...
 
   if (rval) {
     scrub_info();
     scrub_infop->last_local.time = ceph_clock_now(g_ceph_context);
     scrub_infop->last_local.version = get_projected_version();
     scrub_infop->last_scrub_dirty = true;
+  } else if (inode->scrub_infop &&
+	     inode->scrub_infop->header &&
+	     inode->scrub_infop->header->repair) {
+    cache->repair_dirfrag_stats(this, NULL);
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: So we've got this repair function doing real work that can fail (ESTALE), but meanwhile all the callers of scrub_local() are now ignoring its return value. Which means that rstat failures in scrubbing just get lost and never reported. We need to report them somehow (put it in central log and switch to HEALTH_WARN?) while the stats are bad, until they're fixed. (Which may also require some tricky shenanigans when migrations actually happen...not sure how to best handle that. Maybe we can punt on actually verifying stuff gets updated correctly, until multi-MDS is more stable anyway.)

File: src/mds/CInode.cc [hunk 3]
```diff
@@ -3858,22 +3848,28 @@ void CInode::validate_disk_state(CInode::validated_data *results,
...
           ++p) {
-        CDir *dirfrag = in->get_or_open_dirfrag(in->mdcache, *p);
-        if (!dirfrag->is_complete())
-          dirfrag->fetch(gather.new_sub(), false);
+        CDir *dir = in->get_or_open_dirfrag(in->mdcache, *p);
+        if (dir->is_complete()) {
+	  dir->scrub_local();
+	} else {
+	  dir->scrub_info();
+	  dir->scrub_infop->need_scrub_local = true;
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: The whole point of the scrub_info() call is that you can just do
>dir->scrub_info()->need_scrub_local = true;

:)

File: src/mds/CInode.cc [hunk 4]
```diff
@@ -3858,22 +3848,28 @@ void CInode::validate_disk_state(CInode::validated_data *results,
...
       MDSGatherBuilder gather(g_ceph_context);
       std::list<frag_t> frags;
       in->dirfragtree.get_leaves(frags);
       for (list<frag_t>::iterator p = frags.begin();
           p != frags.end();
           ++p) {
-        CDir *dirfrag = in->get_or_open_dirfrag(in->mdcache, *p);
-        if (!dirfrag->is_complete())
-          dirfrag->fetch(gather.new_sub(), false);
+        CDir *dir = in->get_or_open_dirfrag(in->mdcache, *p);
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: Why did you change the name from dirfrag to dir? (I like dirfrag because it disambiguates between the entire [possibly fragmented] directory and an individual frag.)

File: src/mds/CDir.cc [hunk 5]
```diff
@@ -1857,6 +1863,11 @@ void CDir::_omap_fetched(bufferlist& hdrbl, map<string, bufferlist>& omap,
   mark_complete();
   state_clear(STATE_FETCHING);
 
+  if (scrub_infop && scrub_infop->need_scrub_local) {
+    scrub_infop->need_scrub_local = false;
+    scrub_local();
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: We're again losing our chance to report back the results of the local scrub by invoking it as part of the fetch, which will cause trouble for validate_disk_state...

File: src/mds/CInode.cc [hunk 6]
```diff
@@ -3886,47 +3882,54 @@ void CInode::validate_disk_state(CInode::validated_data *results,
...
-          results->raw_rstats.error_str << "dirfrag is INCOMPLETE despite fetching; probably too large compared to MDS cache size?\n";
-          return true;
-        }
-        // FIXME!!! Don't assert out on damage!
-        assert(p->second->scrub_local());
-        sub_info.add(p->second->fnode.accounted_rstat);
+	CDir *dir = p->second;
+	assert(dir->get_version() > 0);
+	nest_info.add(dir->fnode.accounted_rstat);
+	dir_info.add(dir->fnode.accounted_fragstat);
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: ...so we'll report issues above if the dirfrag is gone and fetching returned ENOENT or similar, but not if the frag is somehow damaged (eg, inconsistent rstats). That's not really good enough, we need to actually validate and report the results.

File: src/mds/CInode.cc [hunk 7]
```diff
@@ -3886,47 +3882,54 @@ void CInode::validate_disk_state(CInode::validated_data *results,
 
     bool _dirfrags(int rval) {
       // basic reporting setup
-      results->raw_rstats.checked = true;
-      results->raw_rstats.ondisk_read_retval = rval;
-      results->raw_rstats.passed = false; // we'll set it true if we make it
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: You've dropped initializing this and it needs to start out as false or the validation will never report as failed!

File: src/mds/CDir.cc [hunk 8]
```diff
@@ -2922,8 +2935,6 @@ void CDir::scrub_initialize()
     }
   }
   scrub_infop->directory_scrubbing = true;
-
-  assert(scrub_local()); // TODO: handle failure
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: Why did you decide to drop this one?

File: src/mds/CDir.cc [hunk 9]
```diff
@@ -218,29 +218,43 @@ CDir::CDir(CInode *in, frag_t fg, MDCache *mdcache, bool auth) :
...
-  if(!(get_num_head_items()==
-      (fnode.fragstat.nfiles + fnode.fragstat.nsubdirs))) {
+  if(!frag_info.same_sums(fnode.fragstat)) {
     dout(1) << "mismatch between head items and fnode.fragstat! printing dentries" << dendl;
     dout(1) << "get_num_head_items() = " << get_num_head_items()
              << "; fnode.fragstat.nfiles=" << fnode.fragstat.nfiles
              << " fnode.fragstat.nsubdirs=" << fnode.fragstat.nsubdirs << dendl;
-    for (map_t::iterator i = items.begin(); i != items.end(); ++i) {
-      //if (i->second->get_linkage()->is_primary())
-        dout(1) << *(i->second) << dendl;
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: You've stopped printing out the dentries, despite the error message saying that it will print them all?

File: src/mds/CDir.cc [hunk 10]
```diff
@@ -248,40 +262,32 @@ bool CDir::check_rstats()
...
+  if (!nest_info.same_sums(fnode.rstat)) {
     dout(1) << "mismatch between child accounted_rstats and my rstats!" << dendl;
-    dout(1) << "total of child dentrys: " << sub_info << dendl;
+    dout(1) << "total of child dentrys: " << nest_info << dendl;
     dout(1) << "my rstats:              " << fnode.rstat << dendl;
-    for (map_t::iterator i = items.begin(); i != items.end(); ++i) {
-      if (i->second->get_linkage()->is_primary()) {
-        dout(1) << *(i->second) << " "
-                << i->second->get_linkage()->inode->inode.accounted_rstat
-                << dendl;
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: And also stopped printing out the rstats for each child?

File: src/mds/CDir.cc [hunk 11]
```diff
@@ -218,29 +218,43 @@ CDir::CDir(CInode *in, frag_t fg, MDCache *mdcache, bool auth) :
  * If mds_debug_scatterstat is enabled, assert for correctness,
  * otherwise just print out the mismatch and continue.
  */
-bool CDir::check_rstats()
+bool CDir::check_rstats(bool scrub)
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: This scrub parameter confused me a bit, both in name and purpose. I think it might be clearer if we split it into two different flags. "force" (which forces a check even if mds_debug_scatterstat is disabled), and "assert_failures" which covers whether we assert that everything matches up.

**Comment by gregsfortytwo** [comment]
"mds: properly free CInode::ValidationContinuation" shouldn't be necessary (and I think will break stuff) — the Continuation interface calls delete on _done(). If it's broken we should fix that, not do a deletion here.

File: src/mds/CInode.cc [hunk 12]
```diff
@@ -2136,11 +2139,17 @@ void CInode::finish_scatter_gather_update(int type)
 	    break;
 	  }
 	if (all) {
-	  clog->error() << "unmatched rstat on " << ino() << ", inode has "
-		       << pi->rstat << ", dirfrags have " << rstat << "\n";
-	  assert(!"unmatched rstat" == g_conf->mds_verify_scatter);
+	  if (state_test(CInode::STATE_REPAIRSTATS)) {
+	    dout(20) << " rstat mismatch, fixinga" << dendl;
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: s/fixinga/fixing/

File: src/mds/MDCache.cc [hunk 13]
```diff
@@ -11731,101 +11734,235 @@ void MDCache::scrub_dentry_work(MDRequestRef& mdr)
...
+  le->metablob.add_dir_context(dir);
+  le->metablob.add_dir(dir, true);
+
+  mds->mdlog->submit_entry(le, new C_MDC_RepairDirfragStats(mds, mdr));
+}
+
+void MDCache::repair_inode_stats(CInode *diri, Context *fin)
+{
+  if (!fin)
+    fin = new C_MDSInternalNoop;
```
Comment (gregsfortytwo) [inline src/mds/MDCache.cc]: As with the repair_dirfrag_stats, we aren't doing any reporting of breakage while performing this work, and the only existing caller doesn't pass a Context so the return code is just going to a no-op. Need to fix that.

File: src/mds/CInode.cc [hunk 14]
```diff
@@ -3732,21 +3728,17 @@ void CInode::validate_disk_state(CInode::validated_data *results,
       }
       if (in->is_symlink()) {
         // there's nothing to do for symlinks!
-        results->passed_validation = true;
         return true;
       }
 
-      results->passed_validation = false; // we haven't finished it yet
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: Need to initialize this to false or if we bail out we'll have an undefined result.

File: src/mds/CInode.cc [hunk 15]
```diff
@@ -3765,14 +3757,23 @@ void CInode::validate_disk_state(CInode::validated_data *results,
       // set up basic result reporting and make sure we got the data
       results->performed_validation = true; // at least, some of it!
       results->backtrace.checked = true;
-      results->backtrace.passed = false; // we'll set it true if we make it
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: We still need to start this at false or we'll have an undetermined result if we fail the backtrace check.

File: src/mds/CInode.cc [hunk 16]
```diff
@@ -3765,14 +3757,23 @@ void CInode::validate_disk_state(CInode::validated_data *results,
...
+
+      int64_t pool;
+      if (in->is_dir())
+        pool = in->mdcache->mds->mdsmap->get_metadata_pool();
+      else
+        pool = in->inode.layout.fl_pg_pool;
+      inode_backtrace_t& memory_backtrace = results->backtrace.memory_value;
+      in->build_backtrace(pool, memory_backtrace);
+      bool equivalent, divergent;
+      int memory_newer;
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: There's really no point in pre-declaring equivalent, divergent, and memory_newer.

**Comment by gregsfortytwo** [comment]
Wow, I've just realized we're initializing those passed validation parameters properly in the constructor...ignore those comments!

File: src/mds/CInode.cc [hunk 17]
```diff
@@ -3834,14 +3824,14 @@ void CInode::validate_disk_state(CInode::validated_data *results,
         shadow_in->fetch(get_internal_callback(INODE));
         return false;
       } else {
-        return fetch_dirfrag_rstats();
+	results->inode.passed = true;
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: Why do you want to set inode.passed to true even though we haven't done anything with it? Since we haven't "checked" it, any callers shouldn't even be looking at the "passed" member.

Comment (gregsfortytwo) [inline src/mds/CInode.cc]: ...oh, because you want to do a final check for passed_validation against each of these.

It'd be a little longer but safer to do the passed_validation check by only examining the ones which we checked. Eg
> (!results->raw_stats.checked || results->raw_stats.passed) &&
> (!results->backtrace.checked || results->backtrace.passed) &&
> (!results->inode.checked || results->inode.passed)

How about that instead of confusing the output semantics?

**Comment by gregsfortytwo** [comment]
"mds: set bad dirfrag's version to 1"
A bad dirfrag *is* uninitialized in some ways. If we can't set it to 0, I'm not sure just setting it to 1 is a good solution either — 1 really could be the version it's supposed to be! @jcsp, did you look into any of this when setting up the bad dirfrag code?

File: src/mds/CDir.cc (hunk 8 above)
Comment (ukernel) [inline src/mds/CDir.cc]: because I don't want to call scrub_local() twice for each dirfrag. Once at here, once in CInode::validate_disk_state

**Comment by ukernel** [comment]
On Sat, Jan 23, 2016 at 10:08 AM, Gregory Farnum <notifications@github.com>
wrote:

> "mds: properly free CInode::ValidationContinuation" shouldn't be necessary
> (and I think will break stuff) — the Continuation interface calls delete on
> _done(). If it's broken we should fix that, not do a deletion here.
>
you mean Continuation deletes itself after calling _done()?

File: src/mds/CDir.cc (hunk 8 above)
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: I believe there was actually a purpose to calling it both times — we want to make sure that we aren't skipping any children on setup (so we check the in-memory state), and make sure that we haven't broken anything while the scrub is proceeding (so we check it again at the end). Maybe I'm misremembering and it was just an accident — it's not very expensive though, is it?

**Comment by gregsfortytwo** [comment]
> you mean Continuation deletes itself after calling _done()?

Yes.

**Comment by gregsfortytwo** [comment]
"mds: store ScrubHeader refernce in CInode's scrub_info_t" Why? Are you just making some typing shorter? Because it seems to have increased the size of some parameter lists. (I like changing all the "ScrubHeaderRefConst" params to "const ScrubHeaderRefConst&", though.)

File: src/mds/ScrubStack.h [hunk 18]
```diff
@@ -55,88 +55,77 @@ class ScrubStack {
...
+   * @param in The inodey to scrub
    * @param header The ScrubHeader propagated from whereever this scrub
    *               was initiated
    */
-  void enqueue_dentry_top(CDentry *dn, bool recursive, bool children,
-                          ScrubHeaderRefConst header,
-                          MDSInternalContextBase *on_finish) {
-    enqueue_dentry(dn, recursive, children, header, on_finish, true);
+  void enqueue_inode_top(CInode *in, const ScrubHeaderRefConst& header,
+			 Context *on_finish) {
```
Comment (gregsfortytwo) [inline src/mds/ScrubStack.h]: These changes to take a Context instead of an MDSInternalContextBase are losing us some constraints on the locking associated with those parameters. Is there some change making them safe? It looks like you made the change to accommodate C_MDS_EnqueueScrub, shouldn't the callbacks for admin sockets have a layer between our internal ScrubStack interfaces and reactivating them directly?

**Comment by gregsfortytwo** [comment]
@ukernel, I am realizing as I go through this that you didn't discuss what testing was performed. Did you actually run those ceph-qa-suite tests somewhere? Do something else? Please remember to include that information with all PRs. ;)

File: src/mds/MDCache.cc [hunk 19]
```diff
@@ -11731,101 +11734,235 @@ void MDCache::scrub_dentry_work(MDRequestRef& mdr)
...
+  header->origin = in;
 
-  in->validate_disk_state(vr, mdr, NULL);
+  // only set completion context for non-recursive scrub, because we don't 
+  // want to block asok caller on long running scrub
+  Context *fin = NULL;
+  if (!header->recursive)
+    fin = cs->take_finisher();
+
+  mds->scrubstack->enqueue_inode_bottom(in, header, fin);
```
Comment (gregsfortytwo) [inline src/mds/MDCache.cc]: If this is a non-recursive scrub which we're blocking the admin socket for, we probably want to put it on top of the stack for earliest service, don't we?

File: src/include/ceph_fs.h [hunk 20]
```diff
@@ -350,9 +350,10 @@ enum {
 	// internal op
 	CEPH_MDS_OP_FRAGMENTDIR= 0x01500,
 	CEPH_MDS_OP_EXPORTDIR  = 0x01501,
-	CEPH_MDS_OP_VALIDATE   = 0x01502,
-	CEPH_MDS_OP_FLUSH      = 0x01503,
-	CEPH_MDS_OP_ENQUEUE_SCRUB = 0x01504
+	CEPH_MDS_OP_FLUSH      = 0x01502,
+	CEPH_MDS_OP_ENQUEUE_SCRUB  = 0x01503,
+	CEPH_MDS_OP_REPAIR_FRAGSTATS = 0x01504,
+	CEPH_MDS_OP_REPAIR_INODESTATS = 0x01505
```
Comment (gregsfortytwo) [inline src/include/ceph_fs.h]: Mmm, granted none of these have been exported yet that I can think of, but it's not a good idea to change these numbers since as a group they are shared outside the local daemon. Probably should just change CEPH_MDS_OP_VALIDATE to CEPH_MDS_OP_DEPRECATED and leave the actual numerical values the same.

**Comment by gregsfortytwo** [comment]
"mds: don't ignore scrubbing non-regular non-directory inodes"

We can call validate_disk_state on them, but symlinks still don't get any validation done on them and my recollection is that hard links get resolved to their primary link (even if you come in by way of a secondary link), which is why we skip them in ScrubStack. I'd like to eventually write some code for checking consistency on both of those, but until we do, is there some other reason to call validate_disk_state?

**Comment by gregsfortytwo** [comment]
"mds: make scrubstack queue CInode instead of CDentry" so I agree this can make some things simpler, but the ScrubStack started out on dentries for one very important reason: we want to be able to scrub hard links. If there's some alternate way we can do that, fine, but we'll need a solution.

**Comment by gregsfortytwo** [comment]
And finally, I'm not sure if it never existed or if it got lost somewhere along the way without my noticing, but I'd had the idea that we'd want to be able to scrub a directory and its files without doing a full recursive scrub. That doesn't seem to be possible with the current refer-to-a-shared-ScrubHeader mechanism. Thoughts?

File: src/include/ceph_fs.h (hunk 20 above)
Comment (ukernel) [inline src/include/ceph_fs.h]: they are used in local daemon only

**Comment by ukernel** [comment]
On Tue, Jan 26, 2016 at 1:01 PM, Gregory Farnum <notifications@github.com>
wrote:

> And finally, I'm not sure if it never existed or if it got lost somewhere
> along the way without my noticing, but I'd had the idea that we'd want to
> be able to scrub a directory and its files without doing a full recursive
> scrub. That doesn't seem to be possible with the current
> refer-to-a-shared-ScrubHeader mechanism. Thoughts?
>
> There is some code fragments for that, but never really work.

why refer-to-a-shared-ScrubHeader mechanism does not work (as long as there
is no concurrent scrub)?

**Comment by ukernel** [comment]
On Tue, Jan 26, 2016 at 9:08 AM, Gregory Farnum <notifications@github.com>
wrote:

> "mds: don't ignore scrubbing non-regular non-directory inodes"
>
> We can call validate_disk_state on them, but symlinks still don't get any
> validation done on them and my recollection is that hard links get resolved
> to their primary link (even if you come in by way of a secondary link),
> which is why we skip them in ScrubStack. I'd like to eventually write some
> code for checking consistency on both of those, but until we do, is there
> some other reason to call validate_disk_state?
>
> —
>
because I want to unify "scrub path" and "scrub tag".  the target of "scrub
path" command can be special inode.

**Comment by ukernel** [comment]
On Tue, Jan 26, 2016 at 7:30 AM, Gregory Farnum <notifications@github.com>
wrote:

> "mds: store ScrubHeader refernce in CInode's scrub_info_t" Why? Are you
> just making some typing shorter? Because it seems to have increased the
> size of some parameter lists. (I like changing all the
> "ScrubHeaderRefConst" params to "const ScrubHeaderRefConst&", though.)
>
>
this makes scrub inode code access parameters in ScrubHeader more easily.
It also can be considered as preparation of "mds: make scrubstack queue
CInode instead of CDentry
<https://github.com/ukernel/ceph/commit/eeab3f399da8cb1f9232e47a0e980ac823d09845>
"

File: src/mds/CInode.cc (hunk 3 above)
Comment (ukernel) [inline src/mds/CInode.cc]: On Sat, Jan 23, 2016 at 9:13 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/mds/CInode.cc
> <https://github.com/ceph/ceph/pull/6682#discussion_r50611730>:
>
> >        MDSGatherBuilder gather(g_ceph_context);
> >        std::list<frag_t> frags;
> >        in->dirfragtree.get_leaves(frags);
> >        for (list<frag_t>::iterator p = frags.begin();
> >            p != frags.end();
> >            ++p) {
> > -        CDir *dirfrag = in->get_or_open_dirfrag(in->mdcache, *p);
> > -        if (!dirfrag->is_complete())
> > -          dirfrag->fetch(gather.new_sub(), false);
> > +        CDir *dir = in->get_or_open_dirfrag(in->mdcache, *p);
> > +        if (dir->is_complete()) {
> > +	  dir->scrub_local();
> > +	} else {
> > +	  dir->scrub_info();
> > +	  dir->scrub_infop->need_scrub_local = true;
>
> The whole point of the scrub_info() call is that you can just do
>
> dir->scrub_info()->need_scrub_local = true;
>
>
dir->scrub_info() return const pointer

**Comment by ukernel** [comment]
On Tue, Jan 26, 2016 at 12:59 PM, Gregory Farnum <notifications@github.com>
wrote:

> "mds: make scrubstack queue CInode instead of CDentry" so I agree this can
> make some things simpler, but the ScrubStack started out on dentries for
> one very important reason: we want to be able to scrub hard links. If
> there's some alternate way we can do that, fine, but we'll need a solution.
>
>
scrub remote link it easy, just invoke corresponding function in
ScrubStack::scrub_dirfrag()

File: src/mds/CInode.cc (hunk 16 above)
Comment (ukernel) [inline src/mds/CInode.cc]: On Sat, Jan 23, 2016 at 10:34 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/mds/CInode.cc
> <https://github.com/ceph/ceph/pull/6682#discussion_r50614157>:
>
> > @@ -3765,14 +3757,23 @@ void CInode::validate_disk_state(CInode::validated_data *results,
> >        // set up basic result reporting and make sure we got the data
> >        results->performed_validation = true; // at least, some of it!
> >        results->backtrace.checked = true;
> > -      results->backtrace.passed = false; // we'll set it true if we make it
> > +
> > +      int64_t pool;
> > +      if (in->is_dir())
> > +        pool = in->mdcache->mds->mdsmap->get_metadata_pool();
> > +      else
> > +        pool = in->inode.layout.fl_pg_pool;
> > +      inode_backtrace_t& memory_backtrace = results->backtrace.memory_value;
> > +      in->build_backtrace(pool, memory_backtrace);
> > +      bool equivalent, divergent;
> > +      int memory_newer;
>
> There's really no point in pre-declaring equivalent, divergent, and
> memory_newer.
>

goto is used in this function

File: src/include/ceph_fs.h (hunk 20 above)
Comment (gregsfortytwo) [inline src/include/ceph_fs.h]: I know the ones you're changing are only used locally; but these op IDs in aggregate are shared. It's not a big deal right now I suppose.

**Comment by gregsfortytwo** [comment]
On Tue, Jan 26, 2016 at 4:15 AM, Yan, Zheng <notifications@github.com> wrote:
> why refer-to-a-shared-ScrubHeader mechanism does not work (as long as there
> is no concurrent scrub)?

Well, if we add extra parameters like "scrub_children" it could, but
just as things stand we've got the single (immutable) "recursive"
bool. As long as we're happy with something like that.

**Comment by gregsfortytwo** [comment]
On Wed, Jan 27, 2016 at 2:36 AM, Yan, Zheng <notifications@github.com> wrote:
> scrub remote link it easy, just invoke corresponding function in
> ScrubStack::scrub_dirfrag()

I don't see how that would work. There's no code anywhere that
validates remote links right now, and if somebody (for instance)
passes in a dentry which is itself a hard link, that link gets turned
into an inode and then the inode (in its primary parent location) is
scrubbed — which doesn't help much if the hard link information itself
is somehow stale or is missing the appropriate backlink reference
counts or something.

**Comment by ukernel** [comment]
On Sat, Jan 30, 2016 at 6:28 AM, Gregory Farnum <notifications@github.com>
wrote:

> On Wed, Jan 27, 2016 at 2:36 AM, Yan, Zheng <notifications@github.com>
> wrote:
> > scrub remote link it easy, just invoke corresponding function in
> > ScrubStack::scrub_dirfrag()
>
> I don't see how that would work. There's no code anywhere that
> validates remote links right now, and if somebody (for instance)
> passes in a dentry which is itself a hard link, that link gets turned
> into an inode and then the inode (in its primary parent location) is
> scrubbed — which doesn't help much if the hard link information itself
> is somehow stale or is missing the appropriate backlink reference
> counts or something.


For this case, we can scrub its parent inode. It should be easy to add
 "scrub single dentry in directory" support.

**Comment by gregsfortytwo** [comment]
@jcsp I'd appreciate your briefly looking over the structural changes here and seeing if you have any comments.

File: src/mds/CDir.cc [hunk 21]
```diff
@@ -2974,6 +2994,14 @@ int CDir::_next_dentry_on_set(set<dentry_key_t>& dns, bool missing_okay,
     // okay, we got a  dentry
     dns.erase(dnkey);
 
+    if (dn->get_projected_version() < scrub_infop->last_recursive.version &&
+	!(scrub_infop->header && scrub_infop->header->force)) {
```
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: Can we get here without having set the header? It's a required param of scrub_initialize so I think we're safe.

File: src/mds/CInode.cc [hunk 22]
```diff
@@ -3885,46 +3881,63 @@ void CInode::validate_disk_state(CInode::validated_data *results,
...
-        sub_info.add(p->second->fnode.accounted_rstat);
+	CDir *dir = p->second;
+	assert(dir->get_version() > 0);
+	nest_info.add(dir->fnode.accounted_rstat);
+	dir_info.add(dir->fnode.accounted_fragstat);
+	if (dir->scrub_infop &&
+	    dir->scrub_infop->pending_scrub_error) {
+	  dir->scrub_infop->pending_scrub_error = false;
+	  results->raw_stats.error_str
+	    << "dirfrag(" << p->first << ") has bad stats; ";
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: We might have repaired this, depending on the config setting. The output message should reflect that.

**Comment by jcsp** [comment]
I don't have any objections here, although it looks like not all of Greg's inline comments have responses?

Next time I rebase DamageTable I'll do it on top of this branch.

File: src/mds/CDir.cc (hunk 8 above)
Comment (ukernel) [inline src/mds/CDir.cc]: I still don't add this back in incremental patches. calling  scrub_local() twice for each dirfrag looks weird.

Comment (gregsfortytwo) [inline src/mds/CDir.cc]: Well, it's *very important* we have all the directory's contents in memory when we set up the scrub metadata, because otherwise we'll be skipping children. Obviously that should happen, but scrub_local() was a check to guarantee it. Given the low odds of otherwise detecting missing scrubs of descendants, that's a check I think we need to keep around unless there's a compelling reason not to.

File: src/mds/CDir.cc (hunk 21 above)
Comment (gregsfortytwo) [inline src/mds/CDir.cc]: *bump*

**Comment by gregsfortytwo** [comment]
>For this case, we can scrub its parent inode. It should be easy to add "scrub single dentry in directory" support.

Can you elaborate on how that would work? I'm sure it's possible but, but not certain how to do it cleanly.

File: src/mds/CDir.cc (hunk 21 above)
Comment (ukernel) [inline src/mds/CDir.cc]: On Wed, Feb 17, 2016 at 2:54 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/mds/CDir.cc
> <https://github.com/ceph/ceph/pull/6682#discussion_r53058685>:
>
> > @@ -2974,6 +2994,14 @@ int CDir::_next_dentry_on_set(set<dentry_key_t>& dns, bool missing_okay,
> >      // okay, we got a  dentry
> >      dns.erase(dnkey);
> >
> > +    if (dn->get_projected_version() < scrub_infop->last_recursive.version &&
> > +	!(scrub_infop->header && scrub_infop->header->force)) {
>
> *bump*
>
what do you mean?

File: src/mds/CDir.cc (hunk 8 above)
Comment (ukernel) [inline src/mds/CDir.cc]: On Wed, Feb 17, 2016 at 2:53 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/mds/CDir.cc
> <https://github.com/ceph/ceph/pull/6682#discussion_r53058563>:
>
> > @@ -2922,8 +2935,6 @@ void CDir::scrub_initialize()
> >      }
> >    }
> >    scrub_infop->directory_scrubbing = true;
> > -
> > -  assert(scrub_local()); // TODO: handle failure
>
> Well, it's *very important* we have all the directory's contents in
> memory when we set up the scrub metadata, because otherwise we'll be
> skipping children. Obviously that should happen, but scrub_local() was a
> check to guarantee it. Given the low odds of otherwise detecting missing
> scrubs of descendants, that's a check I think we need to keep around unless
> there's a compelling reason not to.
>
There already is is_complete() check in CDir::scrub_initialize().  Inode's
rstat and fragstat can be wrong, so we can use assert() here. without
assert() I can't see how scrub_local() makes extra guarantee

Comment (ukernel) [inline src/mds/CDir.cc]: On Wed, Feb 17, 2016 at 2:53 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/mds/CDir.cc
> <https://github.com/ceph/ceph/pull/6682#discussion_r53058563>:
>
> > @@ -2922,8 +2935,6 @@ void CDir::scrub_initialize()
> >      }
> >    }
> >    scrub_infop->directory_scrubbing = true;
> > -
> > -  assert(scrub_local()); // TODO: handle failure
>
> Well, it's *very important* we have all the directory's contents in
> memory when we set up the scrub metadata, because otherwise we'll be
> skipping children. Obviously that should happen, but scrub_local() was a
> check to guarantee it. Given the low odds of otherwise detecting missing
> scrubs of descendants, that's a check I think we need to keep around unless
> there's a compelling reason not to.
>

There already is is_complete() check in CDir::scrub_initialize().  Inode's
rstat and fragstat can be wrong, so we can't use assert() here. without
assert() I can't see how scrub_local() makes extra guarantee

**Comment by ukernel** [comment]
On Wed, Feb 17, 2016 at 2:57 AM, Gregory Farnum <notifications@github.com>
wrote:

> For this case, we can scrub its parent inode. It should be easy to add
> "scrub single dentry in directory" support.
>
> Can you elaborate on how that would work? I'm sure it's possible but, but
> not certain how to do it cleanly.
>
something like:

<pre>
diff --git a/src/mds/CDir.cc b/src/mds/CDir.cc
index 91f6dde..bb27a4d 100644
--- a/src/mds/CDir.cc
+++ b/src/mds/CDir.cc
@@ -2930,6 +2930,8 @@ void CDir::scrub_initialize(const
ScrubHeaderRefConst& header)
     // TODO: handle snapshot scrubbing
     if (i->first.snapid != CEPH_NOSNAP)
       continue;
+    if (header->frag == get_frag() && i->first.name != header->dn)
+      continue;

     CDentry::linkage_t *dnl = i->second->get_projected_linkage();
     if (dnl->is_primary()) {
diff --git a/src/mds/CInode.cc b/src/mds/CInode.cc
index 3a41d3d..13bc9ce 100644
--- a/src/mds/CInode.cc
+++ b/src/mds/CInode.cc
@@ -4153,10 +4153,15 @@ void CInode::scrub_initialize(CDentry *scrub_parent,
     for (std::list<frag_t>::iterator i = frags.begin();
         i != frags.end();
         ++i) {
-      if (header->force)
- scrub_infop->dirfrag_stamps[*i].reset();
-      else
- scrub_infop->dirfrag_stamps[*i];
+      if (header->frag != frag_t()) {
+ if (header->frag == *i)
+  scrub_infop->dirfrag_stamps[*i].reset();
+      } else {
+ if (header->force)
+  scrub_infop->dirfrag_stamps[*i].reset();
+ else
+  scrub_infop->dirfrag_stamps[*i];
+      }
     }
   }

diff --git a/src/mds/ScrubHeader.h b/src/mds/ScrubHeader.h
index 72231b5..88623a5 100644
--- a/src/mds/ScrubHeader.h
+++ b/src/mds/ScrubHeader.h
@@ -19,6 +19,11 @@ public:
   bool force;
   bool recursive;
   bool repair;
+
+  // set these two field if you only want to scrub single dentry
+  frag_t frag;
+  string dn;
+
   Formatter *formatter;
 };
 typedef ceph::shared_ptr<ScrubHeader> ScrubHeaderRef;

</pre>

**Comment by gregsfortytwo** [comment]
Failing tests: http://pulpito.ceph.com/gregf-2016-02-17_22:25:54-fs-greg-pr-6682---basic-mira/

**Comment by ukernel** [comment]
tested by vstart_runner.py

**Comment by jcsp** [comment]
Rebased version of this here: https://github.com/jcsp/ceph/tree/wip-scrub-repair

**Comment by ukernel** [comment]
@jscp thanks

re-push jscp's rebase. no other change.

**Comment by jcsp** [comment]
The jenkins failure here is unrelated (http://tracker.ceph.com/issues/15004)

**Comment by gregsfortytwo** [comment]
Tests variously in http://pulpito.ceph.com/gregf-2016-03-07_23:13:46-fs-greg-fs-testing-3-7-damage---basic-mira/, http://pulpito.ceph.com/gregf-2016-03-07_23:15:15-fs-greg-fs-testing-3-7-multifs---basic-mira/, http://pulpito.ceph.com/gregf-2016-03-08_12:17:43-fs-greg-fs-testing-3-7-multifs---basic-smithi/
The failures are known, largely infrastructure issues, and consistent at time of testing.
